internal/powerplatform/services/data_record/api_data_record.go

```diff
@@ -0,0 +1,432 @@
+				delete := true
+				for _, nestedItem := range nestedMapList {
+					nestedMap := nestedItem.(map[string]interface{})
+
+					tableLogicalName, ok := nestedMap["table_logical_name"].(string)
+					if !ok {
+						return nil, fmt.Errorf("table_logical_name field is missing or not a string")
+					}
+					dataRecordId, ok := nestedMap["data_record_id"].(string)
+					if !ok {
+						return nil, fmt.Errorf("data_record_id field is missing or not a string")
+					}
+
+					relationEntityDefinition := getEntityDefinition(ctx, client, environmentUrl, tableLogicalName)
+					if existingRelation.OdataID == fmt.Sprintf("%s/api/data/%s/%s(%s)", environmentUrl, constants.DATAVERSE_API_VERSION, relationEntityDefinition.LogicalCollectionName, dataRecordId) {
+						delete = false
+						break
+					}
+				}
+				if delete {
+					toBeDeleted = append(toBeDeleted, existingRelation)
+				}
+			}
+
+			for _, relation := range toBeDeleted {
+				_, err = client.Api.Execute(ctx, "DELETE", relation.OdataID, nil, nil, []int{http.StatusOK, http.StatusNoContent}, nil)
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			for _, nestedItem := range nestedMapList {
+				nestedMap := nestedItem.(map[string]interface{})
+
+				tableLogicalName, ok := nestedMap["table_logical_name"].(string)
+				if !ok {
+					return nil, fmt.Errorf("table_logical_name field is missing or not a string")
+				}
+				dataRecordId, ok := nestedMap["data_record_id"].(string)
+				if !ok {
+					return nil, fmt.Errorf("data_record_id field is missing or not a string")
+				}
+
+				entityDefinition := getEntityDefinition(ctx, client, environmentUrl, tableLogicalName)
+
+				relation := RelationApiBody{
+					OdataID: fmt.Sprintf("%s/api/data/%s/%s(%s)", environmentUrl, constants.DATAVERSE_API_VERSION, entityDefinition.LogicalCollectionName, dataRecordId),
+				}
+				_, err = client.Api.Execute(ctx, "POST", apiUrl.String(), nil, relation, []int{http.StatusOK, http.StatusNoContent}, nil)
+				if err != nil {
+					return &result, err
+				}
+			}
+		}
+	}
+
+	return &result, nil
+}
+
+func (client *DataRecordClient) DeleteDataRecord(ctx context.Context, recordId string, environmentId string, tableName string, columns map[string]interface{}) error {
+	environmentUrl, err := client.GetEnvironmentUrlById(ctx, environmentId)
+	if err != nil {
+		return err
+	}
+
+	tableEntityDefinition := getEntityDefinition(ctx, client, environmentUrl, tableName)
+
+	e, _ := url.Parse(environmentUrl)
+	apiUrl := &url.URL{
+		Scheme: e.Scheme,
+		Host:   e.Host,
+		Path:   fmt.Sprintf("/api/data/%s/%s(%s)", constants.DATAVERSE_API_VERSION, tableEntityDefinition.LogicalCollectionName, recordId),
+	}
+
+	for key, value := range columns {
+		if _, ok := value.(map[string]interface{}); ok {
+			delete(columns, key)
+		}
+		if nestedMapList, ok := value.([]interface{}); ok {
+			delete(columns, key)
+
+			for _, nestedItem := range nestedMapList {
+				nestedMap := nestedItem.(map[string]interface{})
```

Comment:
again, same code copy/pasted just the result is sent elsewhere. Can you build a helper method that does it and takes func for execution?

Reply:
refactored the code